core-plugins/src/main/java/io/cdap/plugin/batch/joiner/Lookup.java

```diff
@@ -0,0 +1,344 @@
+        failureCollector.addFailure("Missing lookup dataset.", "Lookup dataset must provided.")
+          .withConfigProperty(LOOKUP_DATASET);
+      }
+      if (!containsMacro(INPUT_KEY_FIELD) && Strings.isNullOrEmpty(inputKeyField)) {
+        failureCollector.addFailure("Missing input key field.", "Input key field must be provided.")
+          .withConfigProperty(INPUT_KEY_FIELD);
+      }
+      if (!containsMacro(LOOKUP_KEY_FIELD) && Strings.isNullOrEmpty(lookupKeyField)) {
+        failureCollector.addFailure("Missing lookup key field.", "Lookup key field must be provided.")
+          .withConfigProperty(LOOKUP_KEY_FIELD);
+      }
+      if (!containsMacro(LOOKUP_VALUE_FIELD) && Strings.isNullOrEmpty(lookupValueField)) {
+        failureCollector.addFailure("Missing lookup value field.", "Lookup value field must be provided.")
+          .withConfigProperty(LOOKUP_VALUE_FIELD);
+      }
+      if (!containsMacro(OUTPUT_FIELD) && Strings.isNullOrEmpty(outputField)) {
+        failureCollector.addFailure("Missing output field.", "Output field must be provided.")
+          .withConfigProperty(OUTPUT_FIELD);
+      }
+    }
+  }
+
+  private final Lookup.Config config;
+
+  public Lookup(Config config) {
+    this.config = config;
+  }
+
+  private Schema generateOutputSchema(Map<String, Schema> inputSchemas, FailureCollector collector) {
+    final Set<String> schemas = inputSchemas.keySet();
+    if (schemas.size() == 0) {
+      collector.addFailure("Missing input schema.", "Please connect valid input datasets.");
+    }
+    Schema lookupSchema = inputSchemas.getOrDefault(config.lookupDataset, null);
+    Schema inputSchema = null;
+    if (lookupSchema == null) {
+      collector.addFailure("Lookup dataset with name not found.", "Please provide valid name for lookup dataset.");
+    }
+    for (String schemaName : schemas) {
+      if (!schemaName.equals(config.lookupDataset)) {
+        inputSchema = inputSchemas.getOrDefault(schemaName, null);
+      }
+    }
+    List<Schema.Field> fields = new ArrayList<>(inputSchema.getFields());
+    final Schema.Field lookupField = lookupSchema.getField(config.lookupValueField);
+    fields.add(Schema.Field.of(config.outputField, lookupField.getSchema().isNullable() ? lookupField.getSchema()
+      : Schema.nullableOf(lookupField.getSchema())));
+    return Schema.recordOf("join.output", fields);
+  }
+
+  @Override
+  public void configurePipeline(MultiInputPipelineConfigurer multiInputPipelineConfigurer) {
+    super.configurePipeline(multiInputPipelineConfigurer);
+    final Map<String, Schema> inputSchemas = multiInputPipelineConfigurer
+      .getMultiInputStageConfigurer().getInputSchemas();
+    FailureCollector collector = multiInputPipelineConfigurer.getMultiInputStageConfigurer().getFailureCollector();
+    config.validate(collector);
+    if (config.fieldsContainMacro()) {
+      return;
+    }
+    if (inputSchemas.values().size() != 2) {
```

Comment:
Can you try testing the following:
- 1 input stage's output schema is a macro
- both input stages' output schemas are macros.

I think you'll hit an issue with the former case - please test and see if it works.

If you just need input stage names, you can use `getInputStages`. https://github.com/cdapio/cdap/blob/develop/cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/MultiInputStageConfigurer.java#L44